stuffer/s2n_stuffer_network_order.c

```diff
@@ -0,0 +1,185 @@
+
+    return S2N_SUCCESS;
+}
+
+int s2n_stuffer_read_uint24(struct s2n_stuffer *stuffer, uint32_t * u)
+{
+    uint8_t data[3];
+
+    GUARD(s2n_stuffer_read_bytes(stuffer, data, sizeof(data)));
+
+    *u = data[0] << 16;
+    *u |= data[1] << 8;
+    *u |= data[2];
+
+    return 0;
+}
+
+int s2n_stuffer_write_uint24(struct s2n_stuffer *stuffer, const uint32_t u)
+{
+    return s2n_stuffer_write_network_order(stuffer, u, SIZEOF_UINT24);
+}
+
+int s2n_stuffer_read_uint32(struct s2n_stuffer *stuffer, uint32_t * u)
+{
+    uint8_t data[4];
+
+    GUARD(s2n_stuffer_read_bytes(stuffer, data, sizeof(data)));
+
+    *u = ((uint32_t) data[0]) << 24;
+    *u |= data[1] << 16;
+    *u |= data[2] << 8;
+    *u |= data[3];
+
+    return 0;
+}
+
+int s2n_stuffer_write_uint32(struct s2n_stuffer *stuffer, const uint32_t u)
+{
+    return s2n_stuffer_write_network_order(stuffer, u, sizeof(u));
+}
+
+int s2n_stuffer_read_uint64(struct s2n_stuffer *stuffer, uint64_t * u)
+{
+    uint8_t data[8];
+
+    GUARD(s2n_stuffer_read_bytes(stuffer, data, sizeof(data)));
+
+    *u = ((uint64_t) data[0]) << 56;
+    *u |= ((uint64_t) data[1]) << 48;
+    *u |= ((uint64_t) data[2]) << 40;
+    *u |= ((uint64_t) data[3]) << 32;
+    *u |= ((uint64_t) data[4]) << 24;
+    *u |= ((uint64_t) data[5]) << 16;
+    *u |= ((uint64_t) data[6]) << 8;
+    *u |= data[7];
+
+    return 0;
+}
+
+int s2n_stuffer_write_uint64(struct s2n_stuffer *stuffer, const uint64_t u)
+{
+    GUARD(s2n_stuffer_write_network_order(stuffer, u >> (sizeof(uint32_t) * 8), sizeof(uint32_t)));
+    GUARD(s2n_stuffer_write_network_order(stuffer, u & UINT32_MAX, sizeof(uint32_t)));
+
+    return 0;
+}
+
+static int length_matches_value_check(uint32_t value, uint8_t length)
+{
+    /* Value is represented as a uint32_t, so shouldn't be assumed larger */
+    S2N_ERROR_IF(length > sizeof(uint32_t), S2N_ERR_SIZE_MISMATCH);
+
+    if (length < sizeof(uint32_t)) {
+        /* Value should be less than the maximum for its length */
+        S2N_ERROR_IF(value >= (0x01 << (length * 8)), S2N_ERR_SIZE_MISMATCH);
+    }
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_stuffer_write_reservation(struct s2n_stuffer_reservation reservation, uint32_t u)
+{
+    notnull_check(reservation.stuffer);
+    uint32_t old_write_cursor = reservation.stuffer->write_cursor;
+    reservation.stuffer->write_cursor = reservation.write_cursor;
+
+    if (!s2n_stuffer_is_valid(reservation.stuffer)) {
+        reservation.stuffer->write_cursor = old_write_cursor;
+        S2N_ERROR(S2N_ERR_SAFETY);
+    }
+
+    if (length_matches_value_check(u, reservation.length) != S2N_SUCCESS
+            || s2n_stuffer_write_network_order(reservation.stuffer, u, reservation.length) != S2N_SUCCESS) {
+        reservation.stuffer->write_cursor = old_write_cursor;
+        S2N_ERROR_PRESERVE_ERRNO();
+    }
+
+    reservation.stuffer->write_cursor = old_write_cursor;
+    return S2N_SUCCESS;
+}
```

Comment:
Every exit path from here should restore the `old_write_cursor`.  What if you made that explicit:

```suggestion
static int s2n_stuffer_write_reservation_impl(struct s2n_stuffer_reservation reservation, uint32_t u)
{
    S2N_ERROR_IF(!s2n_stuffer_is_valid(reservation.stuffer), S2N_ERR_SAFETY);
    GUARD(length_matches_value_check(u, reservation.length));
    GUARD(s2n_stuffer_write_network_order(reservation.stuffer, u, reservation.length));
    return S2N_SUCCESS
}

static int s2n_stuffer_write_reservation(struct s2n_stuffer_reservation reservation, uint32_t u)
{
    notnull_check(reservation.stuffer);
    uint32_t old_write_cursor = reservation.stuffer->write_cursor;
    reservation.stuffer->write_cursor = reservation.write_cursor;
    int rval = s2n_stuffer_write_reservation_impl(reservation, u);
    reservation.stuffer->write_cursor = old_write_cursor;
    return rval;
}
```